### ISSUE TITLE
babel throw error when requiring server side code

### ISSUE THREAD

**Issue by abhijeetNmishra** [iheartradio/node-style-loader#8]
I have below code in my React component 

`require('myModule/lib/core.css');`

I am using node-style-loader in my webpack. but it throws following error 

```
(function (exports, require, module, __filename, __dirname) { html{font-family:Targetica,Helvetica Neue,Helvetica,Arial,sans-serif;font-size:15px;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%;box-sizing:border-box}*,:after,:before{box-sizing:inherit}body{line-height:1.42857;color:#333;background-color:#fff;max-width:100%;overflow-x:hidden;margin:0}a{color:inherit;touch-action:manipulation;text-decoration:none}a:focus,a:hover{color:#333}a:focus{outline:1px dashed #666;outline-offset:2px;-moz-outline-radius:4px}article,aside,details,figcaption,figure,footer,header,hgroup,main,menu,nav,section,summary{display:block}audio,canvas,progress,video{display:inline-block;vertical-align:baseline}audio:not([controls]){display:none;height:0}[hidden],template{display:none}a{background-color:transparent}a:active,a:hover{outline:0}abbr[title]{border-bottom:1px dotted}b,strong{font-weight:700}dfn{font-style:italic}h1{font-size:29.29688px}h2{
SyntaxError: Unexpected token {
```

It seems babel is requiring css as js. What is the work around for this.

@iheartradio

**Comment by tarikjn** [comment]
Hi, @abhijeetNmishra what is your loader configuration like?

**Comment by abhijeetNmishra** [comment]
@tarikjn 
 ```
loaders: [{
      test: /\.js$/,
      include: path.join(__dirname, 'client'),
      loaders: ['babel']
    }, {
      test: /\.css$/,
      loader: 'node-style!css-loader'
    }]
```

**Comment by tarikjn** [comment]
This is your server Webpack build right? On the client-side build you should use just `style` instead of `node-style`. Also, you can just write `css` for the `css-loader` since you omitted `-loader` for the rest of your config.

**Comment by abhijeetNmishra** [comment]
@tarikjn do i really need webpack for server side code ? I use babel-node to run my code for Node 6

**Comment by stayman** [comment]
Apologies, I clicked in the wrong window

**Comment by tarikjn** [comment]
If you want your app to be isomorphic/universal and render your views server-side for e.g. first page load speed or SEO, it's a good idea. If you don't pre-render your CSS in that case, you'll get a [FOUC](https://en.wikipedia.org/wiki/Flash_of_unstyled_content). `node-style-loader` is only for the server side part, in your client-side part, you should just use `style-loader`.

**Comment by abhijeetNmishra** [comment]
@tarikjn I am building an universal application, but loader is failing to collect the required css and inlining it with the example mentioned. babel-loader is causing the issue. Also in the documentation, it has not mentioned how to import the css file. Hope i am making myself clear

**Comment by tarikjn** [comment]
@abhijeetNmishra okay I'll update the doc to include loader configuration examples. If you're building an universal application, I strongly urge you to build your server code in Webpack with the same configuration as your client save a few things such as the style-loader. CSS loading will break if you use `babel-node`. Also, `babel-node` is not for production, see: https://babeljs.io/docs/usage/cli/#babel-node. The only thing I would use this for is to run a REPL which has babel support.

**Comment by tarikjn** [comment]
@abhijeetNmishra closing this, please feel free to re-open as needed.

**Comment by therynamo** [comment]
>  Also, you can just write css for the css-loader since you omitted -loader for the rest of your config.

This functionality is deprecated in W2 https://webpack.js.org/guides/migrating/#automatic-loader-module-name-extension-removed

@tarikjn we can run our server code through WP but we've elected not to up until now due to some concerns. 

Though we're writing a universal application, there are still portions of each layer of code that relate to server and client respectively. We don't want to have to run transpiled server code on the server, bare metal is preferred. Do you guys run your server code as transpiled code? If so, do you see any performance differences? Were there tradeoffs that were made in order to enable css loading?

**Comment by stayman** [comment]
@therynamo: I'm not sure exactly what your concerns are around transpiled code on the server, but how we do it is we have two entrypoints, allowing us to separate by layer as necessary. Also, if you're building react or running babel-node you're incurring a performance penalty as described [here](https://babeljs.io/docs/usage/cli/#babel-node) Also, I'm not familiar with a way to import css files into bare metal node.

If you want some idea of how to build a universal app using React/Webpack I recommend [this blog post](https://medium.com/front-end-developers/handcrafting-an-isomorphic-redux-application-with-love-40ada4468af4#.g4sckjfip).

**Comment by therynamo** [comment]
Sorry, I don't think I cleared that up initially, we are not running babel-node in production. We transpile the server code pre-build with babel so it can run on a non-transpiled server. 

Also, yes, thats a wonderful article, I appreciate the link, I've read over it a few times in the past.

So let me rephrase a bit, I think my wording may have caused confusion. We have more of an apprehension than a *concern* about running transpiled server code in production. 

In our eyes it makes developing Node layer code less straight-forward. I understand you can use source-maps to debug your code if you need to, but I think it's about not needing to with so much ES2015 support from Node 6 now. Which leaves us with client side JS and css that need to be transpiled and still rendered from Node. 

This is where my *apprehension* comes from. We *feel* like leaving the purity, for lack of a better word, of the server side code will enable us to avoid incurring issues or bugs with transpiled code that become harder to pin-point when transpiled. Unlike client code that is typically ugli-mini-fied before going to production, as you very well know, server side code doesn't have to incur that process. Which means we would not want to bundle our server side code, each file can live happily in its designated VM somewhere. I've seen some webpack copy plugins, and that may be a good route to go. 

My question was formed from looking at the suggestions here and wondering what your team's process was. Do you use transpiled Node code in production? I haven't looked in some time, but when I did I couldn't find any valuable information regarding the topic on whether people did this or not and the tradeoffs. Though I am curious none the less, maybe my feelings toward this are invalid and I'm simply misunderstanding. 

Thank you for your quick reply @stayman! It is much appreciated.

**Comment by stayman** [comment]
Always happy to respond.

I understand your concerns much better now. And to directly answer your question, we do use transpiled code in production, partially to ensure that we can manage things like non-js dependencies without having to fall back on conditional logic like:

```js
if (process.env.BROWSER === true) {
  styles = require('./foo.css')
} else {
  styles = {};
}
```

in our isomorphic code. I'd be equally interested in other strategies to manage non-js dependencies in the node layer. I'd also point you to [webpack-node-externals](https://www.npmjs.com/package/webpack-node-externals#how-can-i-bundle-required-assets-ie-css-files-from-node_modules) which may be able to be adapted to address your concerns

**Comment by tarikjn** [comment]
I think your concern about debugging transpiled code for the Node side pale in comparison to the penalty that you codebase will take for trying to build an universal app that run through different code pipelines. There are good tools for debugging, and with a cobebase with multiple dozen thousands of LOC and plenty of edge cases and legacy code, we've not encountered such an issue as you describe.

**Comment by therynamo** [comment]
> we do use transpiled code in production

Awesome, @stayman, that is what I was curious about. Did you guys have any process to benchmark any difference between the two? I'm just curious if your team found something that worked well.

>  having to fall back on conditional logic

Yeah, that is an unfortunate trade off.

> I'd also point you to webpack-node-externals

Good call out. I've seen this package mentioned in a few other issues. I'll take a look, thank you!

> I think your concern about debugging transpiled code for the Node side pales in comparison to the penalty that your codebase will take for trying to build an universal app that run through different code pipelines. 

This is a fair statement. I definitely understand there are tradeoffs there. I think that something we'd be interested in is the performance differences, if there are any. I wouldn't imagine that there would be, however when transpiling code that would otherwise run natively, I think it's healthy to be a tad skeptical.

>  we've not encountered such an issue as you describe.

Fair enough, I'd just like to evaluate all options prior to heading down a single direction.

-----------------

Thanks for your discussion guys! You've been more than kind, I appreciate the time you took to respond.

**Comment by stayman** [comment]
@therynamo we haven't benchmarked because we determined that the single stream of code as well as not having to make tradeoffs around asset dependencies (which we leverage heavily) was worth the penalty, if there is one.

Also our focus has been around limiting the difference as much as possible between our client and server side code to reduce cognitive load, so developer convenience factors in as well.
